server/src/main/java/org/opensearch/cluster/routing/remote/InternalRemoteRoutingTableService.java

```diff
@@ -273,6 +326,63 @@ private void uploadIndex(
         }
     }
 
+    private void uploadIndexRoutingDiff(
+        Map<String, Diff<IndexRoutingTable>> indexRoutingTableDiff,
+        String fileName,
+        BlobContainer blobContainer,
+        ActionListener<Void> completionListener
+    ) {
+        BytesReference bytesInput = null;
+        try (BytesStreamOutput streamOutput = new BytesStreamOutput()) {
+            RemoteIndexRoutingTableDiff remoteIndexRoutingTableDiff = new RemoteIndexRoutingTableDiff(indexRoutingTableDiff);
+            remoteIndexRoutingTableDiff.writeTo(streamOutput);
+            bytesInput = streamOutput.bytes();
+        } catch (IOException e) {
+            logger.error(() -> "Failed to serialize IndexRoutingTableDiff: " + e.getMessage(), e);
+            completionListener.onFailure(e);
+            return;
+        }
+
+        if (!(blobContainer instanceof AsyncMultiStreamBlobContainer)) {
+            try {
+                blobContainer.writeBlob(fileName, bytesInput.streamInput(), bytesInput.length(), true);
+                completionListener.onResponse(null);
+            } catch (IOException e) {
+                logger.error(() -> "Failed to write IndexRoutingTableDiff to remote store: " + e.getMessage(), e);
+                completionListener.onFailure(e);
+            }
+            return;
+        }
+
+        try (IndexInput input = new ByteArrayIndexInput("indexroutingdiff", BytesReference.toBytes(bytesInput))) {
+            try (
+                RemoteTransferContainer remoteTransferContainer = new RemoteTransferContainer(
+                    fileName,
+                    fileName,
+                    input.length(),
+                    true,
+                    WritePriority.URGENT,
+                    (size, position) -> new OffsetRangeIndexInputStream(input, size, position),
+                    null,
+                    false
+                )
```

Comment:
Can we reuse RemoteWriteableEntity interfaces?

Reply:
Addressed.